File: sdks/typescript/tests/typecheck-usage.ts
/** Static consumer contract; checked by `npm run typecheck`, not executed. */
import type {
  ArtifactDownload,
  AttestationCapabilities,
  AttestationPublicKey,
  CancellationResult,
  Capabilities,
  RookholdEvent,
  ExecutorOutputEvidence,
  HashedRookholdEvent,
  JobDetail,
  JobAttestationStatus,
  OutputEvidence,
  Receipt,
  EffectiveJobSpec,
  IsolationClass,
  SubmitResult,
  WhoAmI,
} from "../rookhold.js";
import { Rookhold, RookholdError, isolationSatisfies } from "../rookhold.js";
import { Coop, CoopError } from "../coop.js";

const client = new Rookhold("https://rookhold.example", "tenant-key", { timeoutMs: 5_000 });
const legacyClient: Coop = new Coop("https://rookhold.example", "tenant-key");
const legacyError: CoopError = new CoopError("compatibility");
void legacyClient;
void legacyError;

async function consume(): Promise<void> {
  const submitted = await client.submit("python", "print(42)", {
    limits: { wall_seconds: 5, mem_mb: 128 },
    requirements: { minimum_isolation: "linux-shared-kernel" },
    idempotencyKey: "consumer-request-1",
    retryAmbiguous: true,
  });
  const submittedWithHeaders: SubmitResult = await client.submitResult(
    "python",
    "print(42)",
    { idempotencyKey: "consumer-request-2" },
  );
  const detail: JobDetail = await client.get(submitted.job_id);
  const terminal: JobDetail = await client.wait(submitted.job_id);
  const effective: EffectiveJobSpec | null = detail.effective_spec;
  const receipt: Receipt | null = terminal.receipt;
  const executorOutput: ExecutorOutputEvidence | null | undefined =
    receipt?.executor_output;
  const durableOutput: OutputEvidence | undefined = receipt?.output;
  const events: RookholdEvent[] = await client.replay(submitted.job_id);
  const cancellation: CancellationResult = await client.cancelResult(submitted.job_id);
  const cancelledJob: JobDetail["job_id"] | undefined = cancellation.job?.job_id;
  const identity: WhoAmI = await client.whoami();
  const capabilities: Capabilities = await client.capabilities();
  const attestationCapabilities: AttestationCapabilities = capabilities.attestations;
  const attestationStatus: JobAttestationStatus = detail.attestation;
  const attestationTenant: string | null = attestationStatus.tenant;
  const publicKey: AttestationPublicKey = await client.attestationPublicKey();
  const envelope: ArtifactDownload = await client.downloadAttestation(
    submitted.job_id,
  );
  const resultArtifact: ArtifactDownload = await client.downloadResultArtifact(
    submitted.job_id,
    { timeoutMs: 10_000 },
  );
  const observed: IsolationClass = capabilities.execution.isolation_class;
  const postureSatisfied: boolean = isolationSatisfies(
    observed,
    "linux-shared-kernel",
  );
  const concurrentMemory: number = capabilities.limits.concurrent_mem_mb_max;
  if (
    events[0]?.hash_version === 1 &&
    typeof events[0].event_hash === "string"
  ) {
    const hashed = events[0] as HashedRookholdEvent;
    void hashed.event_hash;
  }
  const cancelled: void = await client.cancel(submitted.job_id);
  void [
    effective,
    receipt,
    durableOutput,
    executorOutput,
    events,
    identity,
    submittedWithHeaders,
    postureSatisfied,
    concurrentMemory,
    cancelledJob,
    cancelled,
    attestationCapabilities,
    attestationStatus,
    attestationTenant,
    publicKey,
    envelope,
    resultArtifact,
  ];
}

function preserveAmbiguousSubmission(error: unknown): string | undefined {
  return error instanceof RookholdError ? error.idempotencyKey : undefined;
}

void [consume, preserveAmbiguousSubmission];

File: sdks/python/tests/typecheck_usage.py
"""Static consumer contract; checked by Pyright and Mypy, not executed."""

from typing import Iterator, Optional, cast

from coop import Coop, CoopError
from rookhold import (
    ArtifactDownload,
    AttestationCapabilities,
    AttestationPublicKey,
    CancellationResponse,
    EffectiveJobSpec,
    ExecutionRequirements,
    ExecutorOutputEvidence,
    HashedRookholdEvent,
    IsolationClass,
    JobAttestationStatus,
    JobDetail,
    JobPage,
    JobResult,
    Limits,
    OutputEvidence,
    Receipt,
    Rookhold,
    RookholdEvent,
    RunResult,
    SubmitResponse,
    SubmitResult,
    isolation_satisfies,
)
from rookhold_cli import CliConfig, RookholdCli

client = Rookhold("https://rookhold.example", "tenant-key", timeout=5)
cli_config = CliConfig(
    base_url="https://rookhold.example",
    api_key="tenant-key",
    minimum_isolation="gvisor-application-kernel",
    wait_seconds=60,
    color=False,
    json_output=True,
)
cli_client = RookholdCli(cli_config, client=client)
legacy_client: Coop = Coop("https://rookhold.example", "tenant-key", timeout=5)
legacy_error: CoopError = CoopError("compatibility")
_legacy = (legacy_client, legacy_error)
requirements: ExecutionRequirements = {"minimum_isolation": "linux-shared-kernel"}
submitted: SubmitResponse = client.submit(
    "python",
    "print(42)",
    limits=Limits(wall_seconds=5, mem_mb=128),
    requirements=requirements,
    idempotency_key="consumer-submit-1",
)
submitted_with_metadata: SubmitResult = client.submit_result(
    "python",
    "print(42)",
    requirements=requirements,
    idempotency_key="consumer-submit-2",
)
page: JobPage = client.list(status="running", language="python")
detail: JobDetail = client.get(submitted["job_id"])
attestation_status: JobAttestationStatus = detail["attestation"]
attestation_tenant: Optional[str] = attestation_status["tenant"]
attestation_capabilities: AttestationCapabilities = client.capabilities()[
    "attestations"
]
public_key: AttestationPublicKey = client.attestation_public_key()
envelope: ArtifactDownload = client.download_attestation(submitted["job_id"])
result_artifact: ArtifactDownload = client.download_result_artifact(
    submitted["job_id"], timeout=10
)
receipt: Optional[Receipt] = detail["receipt"]
terminal: JobDetail = client.wait(submitted["job_id"])
events: Iterator[RookholdEvent] = client.stream(submitted["job_id"])
result: JobResult = client.result(submitted["job_id"])
run_result: RunResult = client.run("python", "print(42)")
cancelled: CancellationResponse = client.cancel_result(submitted["job_id"])
hashed: HashedRookholdEvent = cast(HashedRookholdEvent, next(events))
observed_isolation: IsolationClass = "gvisor-application-kernel"
isolation_ok: bool = isolation_satisfies(
    observed_isolation, requirements.get("minimum_isolation", "none")
)

effective: Optional[EffectiveJobSpec] = detail["effective_spec"]
executor_output: Optional[ExecutorOutputEvidence] = (
    receipt.get("executor_output") if receipt is not None else None
)
durable_output: Optional[OutputEvidence] = (
    receipt.get("output") if receipt is not None else None
)

_ = (
    page,
    detail,
    effective,
    receipt,
    durable_output,
    executor_output,
    terminal,
    events,
    result,
    run_result,
    cancelled,
    hashed,
    submitted_with_metadata,
    isolation_ok,
    attestation_status,
    attestation_tenant,
    attestation_capabilities,
    public_key,
    envelope,
    result_artifact,
    cli_client,
)
